### PR TITLE
Implementation of semantic validation flag, pluralize, and multiples

### PR DESCRIPTION
**Pull request by owinterbg**

### CODE REVIEW

**Comment by VisLab** [comment]
I'm confused as to why the HED.xml is part of this pull request.  I thought those had been part of a different one.   Is that because those changes were added after the branch? Alexander should review too.

**Comment by owinterbg** [comment]
it may be because i had used different XMLs to test. Ill send it to
alexander for review.

File: hedvalidation/hedvalidation/hed_dictionary.py [hunk 1]
```diff
@@ -142,16 +140,24 @@ class default units.
...
-        for UNIT_CLASS_DICTIONARY_KEY in HedDictionary.UNIT_CLASS_DICTIONARY_KEYS:
-            unit_class_dictionary = {};
-            unit_class_elements = self._get_elements_by_name(HedDictionary.UNIT_CLASS_ELEMENT);
-            if HedDictionary.UNITS_ELEMENT == UNIT_CLASS_DICTIONARY_KEY:
-                unit_class_dictionary = self._populate_unit_class_units_dictionary(unit_class_elements);
-            elif HedDictionary.DEFAULT_UNITS_FOR_TYPE_ATTRIBUTE == UNIT_CLASS_DICTIONARY_KEY:
-                unit_class_dictionary = self._populate_unit_class_default_unit_dictionary(unit_class_elements);
-            unit_class_dictionaries[UNIT_CLASS_DICTIONARY_KEY] = unit_class_dictionary;
-        return unit_class_dictionaries;
+        unit_class_Elements = self._get_elements_by_name(self.UNIT_CLASS_ELEMENT);
```
Comment (happy5214) [inline hedvalidation/hedvalidation/hed_dictionary.py]: Rename variable to `unit_class_elements`.

File: hedvalidation/hedvalidation/hed_input_reader.py [hunk 2]
```diff
@@ -72,9 +72,11 @@ def __init__(self, hed_input, tag_columns=[2], has_column_names=True, check_for_
         self._tag_columns = self._convert_tag_columns_to_processing_format(tag_columns);
         self._has_column_names = has_column_names;
         self._worksheet_name = worksheet_name;
-        self._hed_dictionary = self._get_hed_dictionary(hed_xml_file);
-        self._tag_validator = TagValidator(self._hed_dictionary, check_for_warnings=check_for_warnings);
+        if run_semantic_validation:
+            self._hed_dictionary = self._get_hed_dictionary(hed_xml_file);
+        self._tag_validator = TagValidator(self._hed_dictionary, check_for_warnings=check_for_warnings, run_semantic_validation=run_semantic_validation);
```
Comment (happy5214) [inline hedvalidation/hedvalidation/hed_input_reader.py]: `self._hed_dictionary` is undefined if semantic validation is off.

File: hedvalidation/hedvalidation/hed_input_reader.py [hunk 3]
```diff
@@ -338,6 +340,7 @@ def _validate_hed_string(self, hed_string):
             hed_string_delimiter = HedStringDelimiter(hed_string);
             validation_issues += self._validate_individual_tags_in_hed_string(hed_string_delimiter);
             validation_issues += self._validate_top_level_in_hed_string(hed_string_delimiter);
+            validation_issues += self._validate_top_level_in_hed_string(hed_string_delimiter);
```
Comment (happy5214) [inline hedvalidation/hedvalidation/hed_input_reader.py]: Duplicate line.

File: hedvalidation/hedvalidation/hed_input_reader.py [hunk 4]
```diff
@@ -367,7 +370,7 @@ def _validate_tag_levels_in_hed_string(self, hed_string_delimiter):
         validation_issues += self._tag_validator.run_tag_level_validators(top_level_tags, formatted_top_level_tags);
         return validation_issues;
 
-    def _validate_top_level_in_hed_string(self, hed_string_delimiter):
+    def _validate_top_level_in_hed_string(self, hed_string_delimiter):  #modify based on validator
```
Comment (happy5214) [inline hedvalidation/hedvalidation/hed_input_reader.py]: Not yet done?

File: hedvalidation/hedvalidation/tag_validator.py [hunk 5]
```diff
@@ -140,12 +147,14 @@ def run_individual_tag_validators(self, original_tag, formatted_tag, previous_or
...
-        validation_issues += self.check_if_tag_is_valid(original_tag, formatted_tag, previous_original_tag,
-                                                        previous_formatted_tag);
-        validation_issues += self.check_if_tag_unit_class_units_are_valid(original_tag, formatted_tag);
-        validation_issues += self.check_if_tag_requires_child(original_tag, formatted_tag);
+        if self._run_semantic_validation:
+            validation_issues += self.check_if_tag_is_valid(original_tag, formatted_tag, previous_original_tag,
+                                                            previous_formatted_tag);
+            validation_issues += self.check_if_tag_unit_class_units_are_valid(original_tag, formatted_tag);
+            validation_issues += self.check_if_tag_requires_child(original_tag, formatted_tag);
         if self._check_for_warnings:
```
Comment (happy5214) [inline hedvalidation/hedvalidation/tag_validator.py]: Move this line and next into semantic validation `if` block.

File: hedvalidation/hedvalidation/tag_validator.py [hunk 6]
```diff
@@ -394,29 +404,75 @@ def check_if_tag_unit_class_units_are_valid(self, original_tag, formatted_tag):
...
+            original_tag_unit_value = self.get_tag_name(original_tag);
             tag_unit_class_units = tuple(self.get_tag_unit_class_units(formatted_tag));
-            if TagValidator.TIME_UNIT_CLASS in tag_unit_classes and TagValidator.is_hh_mm_time(tag_unit_values):
+            if (TagValidator.TIME_UNIT_CLASS in tag_unit_classes
+                    and TagValidator.is_hh_mm_time(formatted_tag_unit_value)):
                 pass;
             elif re.search(TagValidator.DIGIT_EXPRESSION,
-                           TagValidator.strip_off_units_if_valid(tag_unit_values, tag_unit_class_units)):
+                           self.validate_units(original_tag_unit_value,
+                                                         formatted_tag_unit_value,
```
Comment (happy5214) [inline hedvalidation/hedvalidation/tag_validator.py]: Fix indents.

File: hedvalidation/hedvalidation/tag_validator.py [hunk 7]
```diff
@@ -394,29 +404,75 @@ def check_if_tag_unit_class_units_are_valid(self, original_tag, formatted_tag):
...
                 pass
             else:
                 validation_error = error_reporter.report_error_type('unitClass', tag=original_tag,
                                                                     unit_class_units=','.join(tag_unit_class_units));
                 self._increment_issue_count();
         return validation_error;
 
-    @staticmethod
-    def strip_off_units_if_valid(tag_unit_values, tag_unit_class_units):
+    def get_valid_unit_derivative(self, unit):      #change name to get_valid_unit_plural
```
Comment (happy5214) [inline hedvalidation/hedvalidation/tag_validator.py]: Rename to `get_pluralized_unit`.

File: hedvalidation/hedvalidation/tag_validator.py [hunk 8]
```diff
@@ -394,29 +404,75 @@ def check_if_tag_unit_class_units_are_valid(self, original_tag, formatted_tag):
...
+        -------
+        list
+            list of derivative units
+        """
+        derivativeUnits = [unit];
+        if self._hed_dictionary_dictionaries[self.UNIT_SYMBOL_TYPE].get(unit) is None:
+            derivativeUnits.append(pluralize.plural(unit));
+        return derivativeUnits;
+
+    def strip_off_units_if_valid(self, unit_value, unit, is_unit_symbol):
```
Comment (happy5214) [inline hedvalidation/hedvalidation/tag_validator.py]: Add docstring and make private.

**Comment by VisLab** [comment]
Two overall comments:
1)   The style of our existing python code does not conform to standards... (For example semicolons at end of lines, etc.).  Pycharm has orange indicators on right to indicate style violations.  Please review and try to eliminate violations in the various files.

2)  As indicated by Alexander, the XML should be downloaded or provided by the user rather than given as part of the repository (except for testing).

File: hedvalidation/hedvalidation/hed_input_reader.py (hunk 2 above)
Comment (owinterbg) [reply hedvalidation/hedvalidation/hed_input_reader.py]: should the if statement be removed then?

File: hedvalidation/hedvalidation/hed_input_reader.py (hunk 4 above)
Comment (owinterbg) [reply hedvalidation/hedvalidation/hed_input_reader.py]: I don't believe that that was me. Does something seem wrong with the function?

**Pull request merged**